active/0024--message-validation-and-transform.md

```diff
@@ -0,0 +1,162 @@
+
+## Design
+
+Develop a data processing feature which can support:
+
+- Message validation
+- Message transform
+- Message filtering
+
+This data processing step should happen after authorization and before rules engine.
+Probably can make use of the 'message.publish' with a hook priority which
+positions the callback after authorization, but before data bridges.
+
+### High level design
+
+* Introduce 2 different data processing types: `validation` and `tranform`.
+* Make use of the rules engine SQL, but without the `FROM` caluse as it's always the MQTT message being th input.
+
+The each data process rule should have a unique ID. And each rule should have a input topic-filter name which
+can be used to build a topic index for quick look-up.
+
+The common parts can be described as hocon config below:
+
+```
+{
+    id = id1 # unique id assigned by user
+    topic = 't/1' # input
+    sql = "SELECT ....." # processing
+    type = validation | transform
+}
+
+```
+
+### Message Validation
+
+```
+{
+    id = id1
+    topic = 't/1'
+    type = validation
+    action_if_invalid = disconnect # or 'drop'
+    sql = "SELECT schema_decode(my_avro_schema, payload) as decoded WHERE decoded != undefined"
+}
+```
+
+### Message filter
+
+Message filter is just a validation with 'drop' action.
+So there is no need to create a new data process for it.
+
+### Message transform
+
+To support message transform, we would need to create a naming convention for MQTT message construction.
+
+The proposed key word is `new`, by default, the message is copied to `new` then SQL language can
+mutate the `new` in `SELECT` clause.
+
+For example, a rule which can increment `payload.value`.
+
+```
+{
+    id = id2
+    topic = 't/2'
+    type = tranform
+    sql = "SELECT payload.value + 1 as new.payload.value"
+}
+```
+
+To make it easer for user, we should also consider:
+
+- Support all the existing message properties binding, for example `topic`, `qos` `retain`.
+- Use `new.user_properties` for user properties.
+  This is because the user properties in MQTT message object quite deeply nested,
+  so it was not quite easy to use.
+  We are already using `user_properties` keyword in some data bridge actions.
+
+To support message expansion (one transform to many),
+the `new` will be copied as many times as a `FOREACH` statement has to loop.
+
+```
+sql = "FOREACH payload.myarray as element DO element as new.payload
+```
+
+## Configuration Changes
+
+New config which may look like:
+
+```
+message_process = [
+    {
+        description = "drop message if it is not compabitble with my avro schema"
+        topic = 't/1'
+        type = validation
+        action_if_invalid = disconnect
+        sql = "SELECT schema_decode(my_avro_schema, payload) as decoded WHERE decoded != undefined"
+    },
+    {
+        description = "increment payload.value"
+        topic = 't/2'
+        type = tranform
+        sql = "SELECT * as new, payload.value + 1 as new.payload.value"
```

Comment:
My 2 cents. Given that Rule SQL semantics in general provide a DSL to describe how to transform 1 message into 0 or 1 results (or even ≥0 with `FOREACH`), it's not obvious why this type distinction is needed. If running some SQL over a message produces 0 results (or fails to compute) it filters it out in some sense, otherwise transforms. Besides, with this type distinction the SQL syntax in different contexts might need to be constrained (e.g. disallow `FOREACH` in filters, disallow `WHERE` in transforms), which can be unintuitive for the user.

Reply:
do you mean there is no need for 'transform' vs. 'validation' ?
  --- it's mostly for config clarity, e.g. when it's `validation`, there there must be `action_if_invalid`.

or do you mean there is no need for 'transform' vs. 'filter'? 
  --- I don't think 'filter' type is proposed in this doc. "message filter" is just one of the input requirements, it's not mapped to a process type.

Reply:
> do you mean there is no need for 'transform' vs. 'validation' ?

Mostly this. Agree that config clarity is important. I'm just trying to throw some random ideas that could help to _reduce_ the number of new concepts.

Another random thought though: if we want to _disconnect_ the client when it sends garbage, maybe it's the job for the authorization layer?

Reply:
> Another random thought though: if we want to disconnect the client when it sends garbage, maybe it's the job for the authorization layer?

Before this EIP, there were some discussions around:
1. Extend authz checks to support SQL  --- I like this idea very much ;)
2. Extend rule engine to support authz-like actions --- Product team proposed this, and I think @terry-xiaoyu had also came to this according to his comment.

This EIP is kind of striking in the middle because it's an important feature and deserves it's own space.